src/coprocessor/dag/aggr_fn/impl_bit_op.rs

```diff
@@ -0,0 +1,467 @@
+        src_schema: &[FieldType],
+        out_schema: &mut Vec<FieldType>,
+        out_exp: &mut Vec<RpnExpression>,
+    ) -> Result<Box<dyn super::AggrFunction>> {
+        assert_eq!(aggr_def.get_tp(), T::tp());
+
+        // bit operation outputs one column.
+        out_schema.push(
+            FieldTypeBuilder::new()
+                .tp(FieldTypeTp::LongLong)
+                .flag(FieldTypeFlag::UNSIGNED)
+                .build(),
+        );
+
+        // Rewrite expression to insert CAST() if needed.
+        let child = aggr_def.take_children().into_iter().next().unwrap();
+        let mut exp =
+            RpnExpressionBuilder::build_from_expr_tree(child, time_zone, src_schema.len())?;
+        super::util::rewrite_exp_for_bit_op(src_schema, &mut exp).unwrap();
+        out_exp.push(exp);
+
+        Ok(Box::new(AggrFnBitOp::<T>(std::marker::PhantomData)))
+    }
+}
+
+/// The bit operation aggregate functions.
+#[derive(Debug, AggrFunction)]
+#[aggr_function(state = AggrFnStateBitOp::<T>::new())]
+pub struct AggrFnBitOp<T: BitOp>(std::marker::PhantomData<T>);
+
+/// The state of the BitAnd aggregate function.
+#[derive(Debug)]
+pub struct AggrFnStateBitOp<T: BitOp> {
+    c: usize,
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T: BitOp> AggrFnStateBitOp<T> {
+    pub fn new() -> Self {
+        Self {
+            c: T::init_state(),
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<T: BitOp> super::ConcreteAggrFunctionState for AggrFnStateBitOp<T> {
+    type ParameterType = Int;
+
+    #[inline]
+    fn update_concrete(
+        &mut self,
+        _ctx: &mut EvalContext,
+        value: &Option<Self::ParameterType>,
+    ) -> Result<()> {
+        match value {
+            None => Ok(()),
+            Some(value) => {
+                T::op(&mut self.c, *value as usize);
+                Ok(())
+            }
+        }
+    }
+
+    #[inline]
+    fn push_result(&self, _ctx: &mut EvalContext, target: &mut [VectorValue]) -> Result<()> {
+        target[0].push_int(Some(self.c as Int));
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use cop_datatype::EvalType;
+    use tipb_helper::ExprDefBuilder;
+
+    use super::super::AggrFunction;
+    use super::*;
+    use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+    use crate::coprocessor::dag::aggr_fn::parser::AggrDefinitionParser;
+    use cop_datatype::FieldTypeAccessor;
+
+    #[test]
+    fn test_bit_and() {
+        let mut ctx = EvalContext::default();
+        let function = AggrFnBitOp::<BitAnd>(std::marker::PhantomData);
+        let mut state = function.create_state();
+
+        let mut result = [VectorValue::with_capacity(0, EvalType::Int)];
+
+        state.push_result(&mut ctx, &mut result).unwrap();
+        assert_eq!(
+            result[0].as_int_slice(),
+            &[Some(0xffffffffffffffff as u64 as i64)]
```

Comment:
You may write `0xffffffffffffffffu64 as i64`.
(Same below)

Reply:
😕 Then why not just write `-1_i64`

Reply:
Maybe it's to emphasize that the output result of bit ops is actually an unsigned int? I think just `-1` is also ok...

Reply:
@kennytm 0xffffffffffffffff is more clear.

Reply:
@lonng in that case I'd suggest `0xffff_ffff_ffff_ffff_u64` to clearly show that there are 16 `f`s (IIRC Clippy will complain if you don't insert those `_` anyway).